File: tests/test_apps/ruby/comprehensive_test.rb
require "json"
require "tree_sitter_language_pack"

RSpec.describe "TreeSitterLanguagePack comprehensive tests" do
  fixtures_dir = File.expand_path("../fixtures", __dir__)

  describe "process tests" do
    process_fixtures = JSON.parse(File.read(File.join(fixtures_dir, "process.json")))

    process_fixtures.each do |fixture|
      it fixture["name"] do
        config_json = JSON.generate(fixture["config"])
        result_json = TreeSitterLanguagePack.process(fixture["source"], config_json)
        result = JSON.parse(result_json)
        expected = fixture["expected"]

        expect(result["language"]).to eq(expected["language"]) if expected.key?("language")
        if expected.key?("structure_min")
          expect(result["structure"].length).to be >= expected["structure_min"]
        end
        if expected.key?("imports_min")
          expect(result["imports"].length).to be >= expected["imports_min"]
        end
        if expected.key?("error_count")
          expect(result["metrics"]["error_count"]).to eq(expected["error_count"])
        end
      end
    end
  end

  describe "chunking tests" do
    chunking_fixtures = JSON.parse(File.read(File.join(fixtures_dir, "chunking.json")))

    chunking_fixtures.each do |fixture|
      it fixture["name"] do
        config_json = JSON.generate(fixture["config"])
        result_json = TreeSitterLanguagePack.process(fixture["source"], config_json)
        result = JSON.parse(result_json)
        expected = fixture["expected"]

        if expected.key?("chunks_min")
          expect(result["chunks"].length).to be >= expected["chunks_min"]
        end
      end
    end
  end

  describe "setup" do
    it "initializes with multiple languages" do
      config_str = '{"languages":["python","javascript","rust","go","ruby","java","c","cpp"]}'
      expect {
        TreeSitterLanguagePack.init(config_str)
      }.not_to raise_error
    end
  end
end

File: tests/test_apps/go/comprehensive_test.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	tslp "github.com/kreuzberg-dev/tree-sitter-language-pack/packages/go"
)

type ProcessFixture struct {
	Name     string                 `json:"name"`
	Test     string                 `json:"test"`
	Source   string                 `json:"source"`
	Config   map[string]interface{} `json:"config"`
	Expected map[string]interface{} `json:"expected"`
}

func loadProcessFixtures(t *testing.T, filename string) []ProcessFixture {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "fixtures", filename))
	if err != nil {
		t.Fatalf("Failed to read %s: %v", filename, err)
	}
	var fixtures []ProcessFixture
	if err := json.Unmarshal(data, &fixtures); err != nil {
		t.Fatalf("Failed to parse %s: %v", filename, err)
	}
	return fixtures
}

func TestProcessFixtures(t *testing.T) {
	registry, err := tslp.NewRegistry()
	if err != nil {
		t.Fatalf("Failed to create registry: %v", err)
	}
	defer registry.Close()

	fixtures := loadProcessFixtures(t, "process.json")
	for _, fixture := range fixtures {
		t.Run(fixture.Name, func(t *testing.T) {
			// Build a ProcessConfig from the fixture config map
			configJSON, _ := json.Marshal(fixture.Config)
			var config tslp.ProcessConfig
			if err := json.Unmarshal(configJSON, &config); err != nil {
				t.Fatalf("Failed to parse config: %v", err)
			}

			result, err := registry.Process(fixture.Source, config)
			if err != nil {
				t.Fatalf("process() failed: %v", err)
			}

			// Re-serialize the result to a map for flexible assertion checking
			resultJSON, _ := json.Marshal(result)
			var resultMap map[string]interface{}
			if err := json.Unmarshal(resultJSON, &resultMap); err != nil {
				t.Fatalf("Failed to parse result: %v", err)
			}

			// Check metadata sub-fields
			metadata, _ := resultMap["metadata"].(map[string]interface{})

			if lang, ok := fixture.Expected["language"]; ok {
				if metadata["language"] != lang {
					t.Errorf("language = %v, expected %v", metadata["language"], lang)
				}
			}
			if minVal, ok := fixture.Expected["structure_min"]; ok {
				structures, _ := metadata["structure"].([]interface{})
				min := int(minVal.(float64))
				if len(structures) < min {
					t.Errorf("structure count %d < min %d", len(structures), min)
				}
			}
			if minVal, ok := fixture.Expected["imports_min"]; ok {
				imports, _ := metadata["imports"].([]interface{})
				min := int(minVal.(float64))
				if len(imports) < min {
					t.Errorf("imports count %d < min %d", len(imports), min)
				}
			}
		})
	}
}

func TestChunkingFixtures(t *testing.T) {
	registry, err := tslp.NewRegistry()
	if err != nil {
		t.Fatalf("Failed to create registry: %v", err)
	}
	defer registry.Close()

	fixtures := loadProcessFixtures(t, "chunking.json")
	for _, fixture := range fixtures {
		t.Run(fixture.Name, func(t *testing.T) {
			configJSON, _ := json.Marshal(fixture.Config)
			var config tslp.ProcessConfig
			if err := json.Unmarshal(configJSON, &config); err != nil {
				t.Fatalf("Failed to parse config: %v", err)
			}

			result, err := registry.Process(fixture.Source, config)
			if err != nil {
				t.Fatalf("process() failed: %v", err)
			}

			// Re-serialize the result to a map for flexible assertion checking
			resultJSON, _ := json.Marshal(result)
			var resultMap map[string]interface{}
			if err := json.Unmarshal(resultJSON, &resultMap); err != nil {
				t.Fatalf("Failed to parse result: %v", err)
			}

			if minVal, ok := fixture.Expected["chunks_min"]; ok {
				chunks, _ := resultMap["chunks"].([]interface{})
				min := int(minVal.(float64))
				if len(chunks) < min {
					t.Errorf("chunks count %d < min %d", len(chunks), min)
				}
			}
		})
	}
}

func TestSetup(t *testing.T) {
	t.Run("init_with_multiple_languages", func(t *testing.T) {
		configJSON := `{"languages":["python","javascript","rust","go","ruby","java","c","cpp"]}`
		err := tslp.Init(configJSON)
		if err != nil {
			t.Fatalf("Init with multiple languages failed: %v", err)
		}
	})
}
